### PR TITLE
Move updateMenteeData endpoint to Program Controller

### LINKED ISSUE #266: Mentees should be able to change the applied Mentor
**Is your feature request related to a problem? Please describe.**
A mentee can have only one application per program. Mentees should be able to change their primary mentor. But in the current backend, we use `PUT /api/mentors/{id}/mentee` to edit mentee application. But when using that, mentees can't change the appliedMentee because mentorId is in that endpoint as a path paramenter

**Describe the solution you'd like**
Update `PUT /api/mentor/{id}/mentee` endpoint to `PUT /api/program/{id}/mentee`
- Move the `updateMenteeData` method on MentorController.java to ProgramController.java

- Move `updateMenteeData` method on MentorService.java to ProgramService.java
  -  change `mentorId` parameter to `programId`
  - use `menteeRepository.findByProgramIdAndProfileId` to get the optionalMentee object
  - Update the relevant comments and error messages
  - The method should set the assignedMentor too (currently it only updates course,intent,university etc.)

- Move the relavant unit tests

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by pasindur99**
## Purpose

The purpose of this PR is to fix #266 

## Goals

Updating the updateMenteeData endpoint

## Approach

Changed the endpoint path
Updated the unit tests

### Screenshots

N/A

##  Checklist
- [x] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines ( http://bit.ly/sef-best-practices )
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Related PRs

N/A

## Test environment
 
N/A

### CODE REVIEW

File: src/main/java/org/sefglobal/scholarx/service/ProgramService.java [hunk 1]
```diff
@@ -378,4 +378,42 @@ public Mentor getSelectedMentor(long programId, long profileId)
...
+     * @param programId  which is the Program id of the {@link Mentee} to be updated
+     * @param mentee    with the application of the mentee to be updated
+     * @return the updated {@link Mentee}
+     *
+     * @throws ResourceNotFoundException is thrown if the {@link Mentee} doesn't exist
+     * @throws BadRequestException       if the {@link Mentee} is not in the valid state
+     */
+    public Mentee updateMenteeData(long profileId, long programId, Mentee mentee)
+            throws ResourceNotFoundException, BadRequestException {
+        Optional<Mentee> optionalMentee = menteeRepository.findByProgramIdAndProfileId(profileId, programId);
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/ProgramService.java:395]:
```suggestion
        Optional<Mentee> optionalMentee = menteeRepository.findByProgramIdAndProfileId(programId, profileId);
```

File: src/main/java/org/sefglobal/scholarx/service/ProgramService.java [hunk 2]
```diff
@@ -378,4 +378,42 @@ public Mentor getSelectedMentor(long programId, long profileId)
...
+        if (!optionalMentee.isPresent()) {
+            String msg = "Error, Mentee by profile id: " + profileId + " and " +
+                    "mentor id: " + programId + " cannot be updated. " +
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/ProgramService.java:398]:
```suggestion
                    "program id: " + programId + " cannot be updated. " +
```

File: src/main/java/org/sefglobal/scholarx/service/ProgramService.java [hunk 3]
```diff
@@ -378,4 +378,42 @@ public Mentor getSelectedMentor(long programId, long profileId)
...
+                    "Mentee doesn't exist.";
+            log.error(msg);
+            throw new ResourceNotFoundException(msg);
+        }
+
+        Mentee existingMentee = optionalMentee.get();
+        if (EnrolmentState.PENDING.equals(existingMentee.getState())) {
+            existingMentee.setCourse(mentee.getCourse());
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/service/ProgramService.java:406]: Need to set the appliedMentor
```suggestion
            existingMentee.setAppliedMentor(mentee.AppliedMentor());
            existingMentee.setCourse(mentee.getCourse());
```

File: src/test/java/org/sefglobal/scholarx/controller/ProgramControllerTest.java [hunk 4]
```diff
@@ -316,4 +318,32 @@ void getAllMenteesByProgramId_withUnavailableData_thenReturns404() throws Except
...
 	}
+
+	@Test
+	@WithMockUser(username = "user", authorities = {"DEFAULT"})
+	void updateMenteeData_withUnavailableData_thenReturn404() throws Exception {
+		doThrow(ResourceNotFoundException.class)
+				.when(programService)
+				.updateMenteeData(anyLong(), anyLong(), any(Mentee.class));
+
+		mockMvc.perform(put("/api/programs/{mentorId}/mentee", programId)
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/controller/ProgramControllerTest.java:329]:
```suggestion
		mockMvc.perform(put("/api/programs/{programId}/mentee", programId)
```

File: src/test/java/org/sefglobal/scholarx/controller/ProgramControllerTest.java [hunk 5]
```diff
@@ -316,4 +318,32 @@ void getAllMenteesByProgramId_withUnavailableData_thenReturns404() throws Except
...
+	}
+
+	@Test
+	@WithMockUser(username = "user", authorities = {"DEFAULT"})
+	void updateMenteeData_withUnsuitableData_thenReturn400() throws Exception {
+		doThrow(BadRequestException.class)
+				.when(programService)
+				.updateMenteeData(anyLong(), anyLong(), any(Mentee.class));
+
+		mockMvc.perform(put("/api/programs/{mentorId}/mentee", programId)
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/controller/ProgramControllerTest.java:343]:
```suggestion
		mockMvc.perform(put("/api/programs/{programId}/mentee", programId)
```

File: src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java [hunk 6]
```diff
@@ -355,4 +357,49 @@ void getAllMenteesByProgramId_withAvailableData_thenReturnDataFromRepository() t
...
     }
+
+    @Test
+    void updateMenteeData_withValidData_thenReturnUpdatedData()
+            throws ResourceNotFoundException, BadRequestException {
+        mentee.setState(EnrolmentState.PENDING);
+        mentee.setAppliedMentor(mentor);
+        doReturn(Optional.of(mentee))
+                .when(menteeRepository)
+                .findByProfileIdAndAppliedMentorId(anyLong(), anyLong());
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java:368]:
```suggestion
                .findByProgramIdAndProfileId(anyLong(), anyLong());
```

File: src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java [hunk 7]
```diff
@@ -355,4 +357,49 @@ void getAllMenteesByProgramId_withAvailableData_thenReturnDataFromRepository() t
...
+
+        Mentee savedMentee = programService.updateMenteeData(profileId, programId, mentee);
+        assertThat(savedMentee).isNotNull();
+    }
+
+    @Test
+    void updateMenteeData_withUnavailableData_thenThrowResourceNotFound() {
+        doReturn(Optional.empty())
+                .when(menteeRepository)
+                .findByProfileIdAndAppliedMentorId(anyLong(), anyLong());
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java:381]:
```suggestion
                .findByProgramIdAndProfileId(anyLong(), anyLong());
```

File: src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java [hunk 8]
```diff
@@ -355,4 +357,49 @@ void getAllMenteesByProgramId_withAvailableData_thenReturnDataFromRepository() t
...
+
+        Throwable thrown = catchThrowable(
+                () -> programService.updateMenteeData(profileId, programId, mentee));
+        assertThat(thrown)
+                .isInstanceOf(ResourceNotFoundException.class)
+                .hasMessage("Error, Mentee by profile id: 1 and mentor id: 1 cannot be updated. " +
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java:387]:
```suggestion
                .hasMessage("Error, Mentee by profile id: 1 and program id: 1 cannot be updated. " +
```

File: src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java [hunk 9]
```diff
@@ -355,4 +357,49 @@ void getAllMenteesByProgramId_withAvailableData_thenReturnDataFromRepository() t
...
+                        "Mentee doesn't exist.");
+    }
+
+    @Test
+    void updateMenteeData_withUnsuitableData_thenThrowBadRequest() {
+        mentee.setState(EnrolmentState.APPROVED);
+        doReturn(Optional.of(mentee))
+                .when(menteeRepository)
+                .findByProfileIdAndAppliedMentorId(anyLong(), anyLong());
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/service/ProgramServiceTest.java:396]:
```suggestion
                .findByProgramIdAndProfileId(anyLong(), anyLong());
```

File: src/main/java/org/sefglobal/scholarx/controller/ProgramController.java [hunk 10]
```diff
@@ -113,4 +113,16 @@ public Mentor getSelectedMentor(
...
+  public Mentee updateMenteeData(
+          @PathVariable long id,
+          Authentication authentication,
+          @Valid @RequestBody Mentee mentee
+  )
+          throws ResourceNotFoundException, BadRequestException {
+    Profile profile = (Profile) authentication.getPrincipal();
+    return programService.updateMenteeData(profile.getId(), id, mentee);
+  }
 }
```
Comment (Piumal1999) [inline src/main/java/org/sefglobal/scholarx/controller/ProgramController.java:128]: Fix the formatting

**Review by Piumal1999 — CHANGES_REQUESTED** [review]
Good job @pasindur99 
Requested some minor changes

File: src/main/java/org/sefglobal/scholarx/service/ProgramService.java (hunk 3 above)
Comment (pasindur99) [reply src/main/java/org/sefglobal/scholarx/service/ProgramService.java:406]: @Piumal1999  I think this should be `getAppliedMentor()`. Is it?

Comment (Piumal1999) [reply src/main/java/org/sefglobal/scholarx/service/ProgramService.java]: yeah

**Comment by pasindur99** [comment]
Thank you for reviewing the PR.@Piumal1999 
Fixed these minor changes with 0b5bbb3

**Review by Piumal1999 — APPROVED** [review]
Well done @pasindur99 
Thank you

**Comment by Piumal1999** [comment]
The build is failing btw

File: src/test/java/org/sefglobal/scholarx/controller/MentorControllerTest.java [hunk 11]
```diff
@@ -195,34 +195,6 @@ void updateMenteeData_withValidData_thenReturns200() throws Exception {
 				.andExpect(status().isOk());
 	}
```
Comment (Piumal1999) [inline src/test/java/org/sefglobal/scholarx/controller/MentorControllerTest.java:196]: We need to move this one also

**Review by Piumal1999 — CHANGES_REQUESTED** [review]

File: src/test/java/org/sefglobal/scholarx/controller/MentorControllerTest.java (hunk 11 above)
Comment (pasindur99) [reply src/test/java/org/sefglobal/scholarx/controller/MentorControllerTest.java:196]: I'll fix it ASAP.

**Review by Piumal1999 — APPROVED** [review]
Awesome. Thank you

**Pull request merged**